iml-agent/src/daemon_plugins/ntp.rs

```diff
@@ -0,0 +1,340 @@
+    });
+    let time_offset = get_chronyc_ntpdata_command().map_ok(parse_chrony_time_offset);
+
+    future::try_join(ntp_synced, time_offset).await
+}
+
+async fn time_synced<
+    F1: Future<Output = Result<bool, ImlAgentError>>,
+    F2: Future<Output = Result<(RunState, RunState), ImlAgentError>>,
+    F3: Future<Output = Result<(Option<time::Synced>, Option<time::Offset>), ImlAgentError>>,
+    F4: Future<Output = Result<(Option<time::Synced>, Option<time::Offset>), ImlAgentError>>,
+>(
+    is_ntp_configured_by_iml: fn() -> F1,
+    get_time_sync_services: fn() -> F2,
+    ntpd_synced: fn() -> F3,
+    chronyd_synced: fn() -> F4,
+) -> Result<Output, ImlAgentError> {
+    let configured = is_ntp_configured_by_iml().await?;
+
+    tracing::debug!("Configured: {:?}", configured);
+
+    let (ntp_runstate, chrony_runstate) = get_time_sync_services().await?;
+
+    tracing::debug!(
+        "ntp runstate: {:?}. chrony runstate: {:?}",
+        ntp_runstate,
+        chrony_runstate
+    );
+
+    let state = if ntp_runstate >= RunState::Started && chrony_runstate >= RunState::Started {
+        // Both chronyd and ntpd should not be running at the same time.
+        time::State::Multiple
+    } else if ntp_runstate < RunState::Started && chrony_runstate < RunState::Started {
+        // Neither chronyd or ntpd are running
+        time::State::None
+    } else if ntp_runstate >= RunState::Started {
+        let (ntp_synced, ntp_offset) = ntpd_synced().await?;
+
+        match ntp_synced {
+            Some(time::Synced::Synced) => time::State::Synced,
+            Some(time::Synced::Unsynced) => time::State::Unsynced(ntp_offset),
+            None => time::State::Unknown,
+        }
+    } else {
+        let (chrony_synced, chrony_offset) = chronyd_synced().await?;
+
+        match chrony_synced {
+            Some(time::Synced::Synced) => time::State::Synced,
+            Some(time::Synced::Unsynced) => time::State::Unsynced(chrony_offset),
+            None => time::State::Unknown,
+        }
+    };
+
+    let x = serde_json::to_value(state).map(Some)?;
+
+    Ok(x)
+}
+
+impl DaemonPlugin for Ntp {
+    fn start_session(
+        &mut self,
+    ) -> Pin<Box<dyn Future<Output = Result<Output, ImlAgentError>> + Send>> {
+        self.update_session()
+    }
+
+    fn update_session(
+        &self,
+    ) -> Pin<Box<dyn Future<Output = Result<Output, ImlAgentError>> + Send>> {
+        Box::pin(time_synced(
+            is_ntp_configured_by_iml,
+            get_time_sync_services,
+            ntpd_synced,
+            chronyd_synced,
+        ))
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use futures::FutureExt;
+
+    #[test]
+    fn test_is_ntp_synced() {
+        let s = r#"b true"#;
+
+        assert_eq!(parse_ntp_synced(s), Some(time::Synced::Synced));
+    }
+
+    #[test]
+    fn test_is_ntp_not_synced() {
+        let s = r#"b false"#;
+
+        assert_eq!(parse_ntp_synced(s), Some(time::Synced::Unsynced));
```

Comment:
These tests are tautological. `parse_ntp_synced` is literally mapping strings to an enum.

Reply:
I didn't write these tests, but I guess the question is whether we gain anything by enforcing this invariant. 

Likewise does it harm anything to leave them here?

Reply:
No harm other than compile and execution time. And if we delete them as soon as they fail :)